text/0000-scoped-custom-elements.md

```diff
@@ -0,0 +1,109 @@
+3. Since packages uses custom registries there is no chance for name collision or a component hijacking.
+
+## Detailed design
+
+An LWC package defines own custom elements registry and keeps the reference to be passed to the components.
+
+`registry.js`
+
+```javascript
+export const registry = new CustomElementRegistry();
+```
+
+The custom element used in an LWC has to support a way of passing the registry reference to the shadow root creation. This would require the included custom element to support a static property on the class that keeps the reference to the registry. When the element is being created it uses this reference to pass it to the shadow root creation.
+
+`@scope/custom-element/src/CustomElement.js`
+
+```javascript
+export declare class CustomElement extends HTMLElement {
+  constructor() {
+    const registry = CustomElement.registry;
+    this.attachShadow({mode: 'open', registry});
+  }
+}
+```
+
+The LWC component imports the custom element, passes the handler to the custom element, and registers it.
+
+```javascript
+import { LightningElement } from 'lwc';
+import { CustomElement } from '@scope/custom-element';
+import { registry } from './registry.js';
+CustomElement.registry = registry;
+registry.define('custom-element', CustomElement);
+
+export default class Example extends LightningElement {
+  ...
+}
+```
+
+```html
+<template>
+  <custom-element></custom-element>
+</template>
+```
+
+The whole registration process of external custom elements could be done in the `registry.js` file instead of inside an LWC component to avoid name duplicates.
+
+Finally, the LWC component uses the same registry to register itself in the same registry as the rest of the components.
+
+### Polyfills
+
+There is [Scoped CustomElementRegistry polyfill](https://github.com/webcomponents/polyfills/tree/master/packages/scoped-custom-element-registry) that can be used as a base for the implementation. However, it is a work in progress.
+
+## Drawbacks
+
+Why should we *not* do this? Please consider:
+
+- implementation cost, both in term of code size and complexity
+- whether the proposed feature can be implemented in user space
+- the impact on teaching people the proposed feature
+- integration of this feature with other existing and planned features
+- cost of migrating existing applications (is it a breaking change?)
+
+I see a problem with the static field with the registry on the element. Generally I hate an idea of having static fields to pass references to a crucial part of the registration process. Even though it can't be altered by another application, an author can easily override it with another registry. This could be potentially beneficial in cases when an LWC component is made to be sharable so this component's registry is also dynamic. But it could potentially mean that the scope of the registration change without author realizing what is happening.
```

Comment:
There is a security issue here as well... that static property in LWC is publicly exposed, anyone importing a component will immediate have access to the internals (other custom elements) that otherwise they should not have access to.